llvm-hs-pure/src/LLVM/AST/Typed.hs

```
@@ -0,0 +1,125 @@
+  typeOf (C.GlobalReference t _) = t
+  typeOf (C.Add {..})     = typeOf operand0
+  typeOf (C.FAdd {..})    = typeOf operand0
+  typeOf (C.FDiv {..})    = typeOf operand0
+  typeOf (C.FRem {..})    = typeOf operand0
+  typeOf (C.Sub {..})     = typeOf operand0
+  typeOf (C.FSub {..})    = typeOf operand0
+  typeOf (C.Mul {..})     = typeOf operand0
+  typeOf (C.FMul {..})    = typeOf operand0
+  typeOf (C.UDiv {..})    = typeOf operand0
+  typeOf (C.SDiv {..})    = typeOf operand0
+  typeOf (C.URem {..})    = typeOf operand0
+  typeOf (C.SRem {..})    = typeOf operand0
+  typeOf (C.Shl {..})     = typeOf operand0
+  typeOf (C.LShr {..})    = typeOf operand0
+  typeOf (C.AShr {..})    = typeOf operand0
+  typeOf (C.And {..})     = typeOf operand0
+  typeOf (C.Or  {..})     = typeOf operand0
+  typeOf (C.Xor {..})     = typeOf operand0
+  typeOf (C.GetElementPtr {..}) = getElementPtrType (typeOf address) indices
+  typeOf (C.Trunc {..})   = type'
+  typeOf (C.ZExt {..})    = type'
+  typeOf (C.SExt {..})    = type'
+  typeOf (C.FPToUI {..})  = type'
+  typeOf (C.FPToSI {..})  = type'
+  typeOf (C.UIToFP {..})  = type'
+  typeOf (C.SIToFP {..})  = type'
+  typeOf (C.FPTrunc {..}) = type'
+  typeOf (C.FPExt {..})   = type'
+  typeOf (C.PtrToInt {..}) = type'
+  typeOf (C.IntToPtr {..}) = type'
+  typeOf (C.BitCast {..})  = type'
+  typeOf (C.ICmp {..})    = case (typeOf operand0) of
+                              (VectorType n _) -> VectorType n i1
+                              _ -> i1
+  typeOf (C.FCmp {..})    = case (typeOf operand0) of
+                              (VectorType n _) -> VectorType n i1
+                              _ -> i1
+  typeOf (C.Select {..})  = typeOf trueValue
+  typeOf (C.ExtractElement {..})  = case typeOf vector of
+                                      (VectorType _ t) -> t
+                                      _ -> VoidType {- error "The first operand of an extractelement instruction is a value of vector type." -}
```

Afaik, `ExtractElement` (and the same applies to `ShuffleVector`) really only work for vector types so I think `error` is a better solution than silently ignoring the problem and using `VoidType`.

---

Makes sense, it should error out on any malformed ASTs now rather than swallowing them as `Void`.